File: lab1/task_2/classes/JewelryShop.cpp
#include "Jewelry.cpp"
#include <iostream>

using namespace std;

class JewelryShop {
  private:
    Jewelry jewelry1;
    Jewelry jewelry2;
    Jewelry jewelry3;
    int count1;
    int count2;
    int count3;
    double extraPrice;
    char name[100];

  public:
    void init(char name[], Jewelry j1, int c1, Jewelry j2, int c2, Jewelry j3, int c3, double arg_extraPrice);

    void read();

    void display();

    double getFullPrice();

    Jewelry mostExpensiveJewelry();
};

void JewelryShop::init(char arg_name[], Jewelry j1, int c1, Jewelry j2, int c2, Jewelry j3, int c3, double arg_extraPrice) {
      jewelry1 = j1;
      jewelry2 = j2;
      jewelry3 = j3;

      count1 = c1;
      count2 = c2;
      count3 = c3;

      strcpy(name, arg_name);

      extraPrice = arg_extraPrice;
    };

    void JewelryShop::read() {
      cout << "Введите название магазина: ";
      cin >> name;

      cout << "Введите количество украшений 1: ";
      cin >> count1;
      jewelry1.read();

      cout << "Введите количество украшений 2: ";
      cin >> count2;
      jewelry2.read();

      cout << "Введите количество украшений 3: ";
      cin >> count3;
      jewelry3.read();

      cout << "Введите стоимость доп. аксессуаров: ";
      cin >> extraPrice;
    };

void JewelryShop::display() {
  cout << "Название магазина: " << name << endl;

  cout << "Украшение 1: " << endl;
  jewelry1.display();
  cout << "Количество: " << count1 << endl;

  cout << "Украшение 2: " << endl;
  jewelry2.display();
  cout << "Количество: " << count2 << endl;

  cout << "Украшение 3: " << endl;
  jewelry3.display();
  cout << "Количество: " << count3 << endl;

  cout << "Стоимость доп. аксессуаров: " << extraPrice << endl;
};

double JewelryShop::getFullPrice() {
  return jewelry1.getFullPricePerGramm() * count1 + jewelry2.getFullPricePerGramm() * count2 + jewelry3.getFullPricePerGramm() * count3 + extraPrice;
};

Jewelry JewelryShop::mostExpensiveJewelry() {
  double price1 = jewelry1.getFullPricePerGramm();
  double price2 = jewelry2.getFullPricePerGramm();
  double price3 = jewelry3.getFullPricePerGramm();

  if (price1 > price2 && price1 > price3) {
    return jewelry1;
  } else if (price2 > price1 && price2 > price3) {
    return jewelry2;
  } 
  return jewelry3;
};

File: part_2/lab1/patterns/JewelryShop.cs
using System;

namespace lab1
{
    public class JewelryShop
    {
        private Jewelry jewelry1;
        private Jewelry jewelry2;
        private Jewelry jewelry3;
        private int count1;
        private int count2;
        private int count3;

        public double ExtraPrice { get; set; }

        private string name;

        public void Init(string arg_name, Jewelry j1, int c1, Jewelry j2, int c2, Jewelry j3, int c3, double arg_extraPrice)
        {
            jewelry1 = j1;
            jewelry2 = j2;
            jewelry3 = j3;

            count1 = c1;
            count2 = c2;
            count3 = c3;

            name = arg_name;
            ExtraPrice = arg_extraPrice;
        }

        public void Read()
        {
            Console.Write("Введите название магазина: ");
            name = Console.ReadLine();

            Console.Write("Введите количество украшений 1: ");
            count1 = Convert.ToInt32(Console.ReadLine());
            jewelry1 = new Jewelry();
            jewelry1.Read();

            Console.Write("Введите количество украшений 2: ");
            count2 = Convert.ToInt32(Console.ReadLine());
            jewelry2 = new Jewelry();
            jewelry2.Read();

            Console.Write("Введите количество украшений 3: ");
            count3 = Convert.ToInt32(Console.ReadLine());
            jewelry3 = new Jewelry();
            jewelry3.Read();

            Console.Write("Введите стоимость доп. аксессуаров: ");
            ExtraPrice = Convert.ToDouble(Console.ReadLine());
        }

        public void Display()
        {
            Console.WriteLine($"Название магазина: {name}");

            Console.WriteLine("Украшение 1:");
            jewelry1.Display();
            Console.WriteLine($"Количество: {count1}");

            Console.WriteLine("Украшение 2:");
            jewelry2.Display();
            Console.WriteLine($"Количество: {count2}");

            Console.WriteLine("Украшение 3:");
            jewelry3.Display();
            Console.WriteLine($"Количество: {count3}");

            Console.WriteLine($"Стоимость доп. аксессуаров: {ExtraPrice}");
        }

        public double GetFullPrice()
        {
            return jewelry1.GetFullPricePerGramm() * count1 +
                   jewelry2.GetFullPricePerGramm() * count2 +
                   jewelry3.GetFullPricePerGramm() * count3 +
                   ExtraPrice;
        }

        public Jewelry MostExpensiveJewelry()
        {
            double price1 = jewelry1.GetFullPricePerGramm();
            double price2 = jewelry2.GetFullPricePerGramm();
            double price3 = jewelry3.GetFullPricePerGramm();

            if (price1 > price2 && price1 > price3)
            {
                return jewelry1;
            }
            else if (price2 > price1 && price2 > price3)
            {
                return jewelry2;
            }
            return jewelry3;
        }
    }
}
